feat(curriculum): add merge sort project to Python curriculum

ihechikara opened the pull request:
- break down project into steps
- add step 1 - 28

Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any additional description of changes below this line -->

ihechikara committed:
feat(curriculum): add merge sort project to Python curriculum

- break down project into steps
- add step 1 - 28

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/655cd899f8de09431eabb40c.md:
```diff
@@ -0,0 +1,30 @@
+---
+id: 655cd899f8de09431eabb40c
+title: Step 1
+challengeType: 0
```
```suggestion
challengeType: 20
```

gikf commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569b20f829b7e69d43c232a.md:
```diff
@@ -0,0 +1,55 @@
+
+The last thing to do for the `while` loop is to move to the next index in the sorted array.
+
+Using the increment operator, add 1 to the value of `sorted_index`
```
```suggestion
Using the augmented assignment addition operator, add 1 to the value of `sorted_index`
```

ihechikara committed:
start writing tests

ihechikara committed:
feat: add tests and update PR

- write tests for step 1 to 31 of merge sort project
- update descriptions with PR review suggestions

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/655cd899f8de09431eabb40c.md:
```diff
@@ -0,0 +1,35 @@
+
+# --description--
+
+In this project, you'll learn data structures by building the merge sort algorithm. This is a sorting algotithm that uses the divide-and-conquer principle to sort collections of data. 
```
```suggestion
In this project, you'll learn data structures by building the merge sort algorithm. This is a sorting algorithm that uses the divide-and-conquer principle to sort collections of data. 
```

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/655cd899f8de09431eabb40c.md:
```diff
@@ -0,0 +1,35 @@
+
+In this project, you'll learn data structures by building the merge sort algorithm. This is a sorting algotithm that uses the divide-and-conquer principle to sort collections of data. 
+
+Create a function called `merge_sort`. This function will handle the task of sorting an list of numbers. 
```
```suggestion
Create a function called `merge_sort`. This function will handle the task of sorting a list of numbers. 
```

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6564a9fe51964c229d5b7f4c.md:
```diff
@@ -0,0 +1,33 @@
+
+# --description--
+
+We'll need a parameter that denotes the data collection to be sorted. Create a parameter called `array` in the `merge_sort` function.
```
```suggestion
You'll need a parameter that denotes the data collection to be sorted. Create a parameter called `array` in the `merge_sort` function.
```

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6564aee9c077774ea49c3faf.md:
````diff
@@ -0,0 +1,72 @@
+})
+```
+
+There should be no `pass` keyword in the body of the function. `middle_point = len(array) // 2` should be on the same line where you removed the `pass` keyword
````
Both hints are same.

ihechikara replied:
Yeah. During testing, the second test was working but I was getting the hints message for the first test. So I modified the hints to target both tests. 

This also applies to Step 12

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656627b47bd2d2a4afbc945d.md:
```diff
@@ -0,0 +1,62 @@
+
+# --description--
+
+In the previous step, we got the mid point. You can use it to divide `array` into two and assign each part to new variables. 
```
Slicing is already discussed in Luhn's algorithm. Maybe this instruction can be shortened a bit.

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6577320da0d4c2e594d418e2.md:
```diff
@@ -0,0 +1,43 @@
+
+# --hints--
+
+`right_part` should be a parameter of the function. Paremeters are defined within the parenthesis
```
```suggestion
`right_part` should be a parameter of the function. Parameters are defined within the parenthesis.
```

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656639771fed09b5c0e8fe71.md:
```diff
@@ -0,0 +1,49 @@
+
+# --description--
+
+Now that we've divided the `array` list into two separate lists, we'll do the same thing for each list until every element stands alone.
```
```suggestion
Now that you've divided the `array` list into two separate lists, you'll do the same thing for each list until every element stands alone.
```

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656639771fed09b5c0e8fe71.md:
```diff
@@ -0,0 +1,49 @@
+
+The idea is to keep dividing the array until individual elements are considered. A list with a single number is always sorted.
+
+To do that, we have to recursively call the `merge_sort` function on the `left_part` and `right_part`.
```
```suggestion
To do that, you have to recursively call the `merge_sort` function on the `left_part` and `right_part`.
```

zairahira reviewed (commented):
Reviewed step 1-10.

zairahira reviewed (commented):
Reviewed till step 20.

zairahira commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569beee367427c90c74899e.md:
```diff
@@ -0,0 +1,64 @@
+        right_array_index += 1
+        sorted_index += 1
+
+if __name__ == '__main__':
```
It would be helpful to briefly dd some context about `__main__`.

zairahira reviewed (commented):
Reviewed all the steps :) !

ihechikara committed:
implement PR review suggestions

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md:
```diff
@@ -0,0 +1,76 @@
+
+# --description--
+
+At this point, the `numbers` list has been sorted. Let's print the sorted list.
+
+Start by creating a `print()` function that prints string `Sorted array: ` and the the sorted list.
```
```suggestion
At this point, the `numbers` list has been sorted. Call the `print` function to print string `Sorted array: ` and the sorted list.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md:
```diff
@@ -0,0 +1,76 @@
+
+Start by creating a `print()` function that prints string `Sorted array: ` and the the sorted list.
+
+You can do this by concatenating `"Sorted array: "` and `str(numbers)` in the `print` function.
+
+The `str` function converts the list to a string.
```
```suggestion
To do that, concatenate `"Sorted array: "` and `str(numbers)` in the `print` function.

With that, the merge sort algorithm is complete.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md:
```diff
@@ -0,0 +1,76 @@
+
+# --hints--
+
+You should have `print("sorted array: " + str(numbers))`. Remember to add the space after the colon
```
```suggestion
You should have `print("sorted array: " + str(numbers))` in your code. Remember to add the space after the colon.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md:
```diff
@@ -0,0 +1,76 @@
+        right_array_index += 1
+        sorted_index += 1
+
+if __name__ == '__main__':
+    numbers = [4, 10, 6, 14, 2, 1, 8, 5]
+
+    print("Unsorted array: ")
+    print(numbers)
+
+    merge_sort(numbers)
+
+--fcc-editable-region--
+
+--fcc-editable-region--
```
```suggestion
--fcc-editable-region--
if __name__ == '__main__':
    numbers = [4, 10, 6, 14, 2, 1, 8, 5]
    print("Unsorted array: ")
    print(numbers)    
    merge_sort(numbers)

--fcc-editable-region--
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md:
````diff
@@ -0,0 +1,76 @@
+--fcc-editable-region--
+
+--fcc-editable-region--
+```
````
Please add the solution. See this as a reference: https://github.com/freeCodeCamp/freeCodeCamp/blob/4a365edfa6890afbdaacb2bc3b9ac38a0ed418c8/curriculum/challenges/english/20-upcoming-python/learn-recursion-by-solving-the-tower-of-hanoi-puzzle/657b6a4a49faec5c600287ba.md?plain=1#L58-L84

ihechikara committed:
update PR

ihechikara commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656680b0fc79f2c38a34d90e.md:
```diff
@@ -0,0 +1,122 @@
+Create three variables: `left_array_index`, `right_array_index`, and `sorted_index`. Set their values to be `0`.
+
+# --hints--
+
```
I'll keep this in mind for later.

ihechikara commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656702f8b4cbd8cbf0a433c6.md:
````diff
@@ -0,0 +1,65 @@
+})
+```
+
+Use the `pass` keyword in the body of the loop
+
+```js
+({
+  test: () => {
+    const transformedCode = e.code.original["main.py"].replace(/\r/g, "");
+    const merge_sort = __helpers.python.getDef("\n" + transformedCode, "merge_sort");
+    const { function_body } = merge_sort;
+
+    assert.match(function_body, /while\s+left_array_index\s*<\s*len\(left_part\)\s+and\s+right_array_index\s*<\s*len\(right_part\):\s*pass/);
+  }
+})
+```
````
This checks that `pass` exists. Nothing should be in the body of the function at this point except the `pass` keyword.

ihechikara commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656706afd65547d22bee0b68.md:
```diff
@@ -0,0 +1,57 @@
+challengeType: 20
+dashedName: step-11
+---
+
```
Fixed that.

Dario-DC commented on curriculum/challenges/_meta/learn-data-structures-by-building-the-merge-sort-algorithm/meta.json:
```diff
@@ -0,0 +1,149 @@
+  "usesMultifileEditor": true,
+  "hasEditableBoundaries": true,
+  "dashedName": "learn-data-structures-by-building-the-merge-sort-algorithm",
+  "order": 42,
```
```suggestion
  "order": 2,
```
Change this to please the tests

ihechikara replied:
Done

ihechikara committed:
change meta order

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656627b47bd2d2a4afbc945d.md:
````diff
@@ -0,0 +1,57 @@
+## --seed-contents--
+
+```py
+def merge_sort(array):
+    
+    middle_point = len(array) // 2
+
+--fcc-editable-region--
+
+--fcc-editable-region--
````
```suggestion

--fcc-editable-region--
def merge_sort(array):
    pass
--fcc-editable-region--
```

Dario-DC replied:
```suggestion
--fcc-editable-region--
def merge_sort(array):    
    middle_point = len(array) // 2
--fcc-editable-region--
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6564aee9c077774ea49c3faf.md:
```diff
@@ -0,0 +1,53 @@
+    const merge_sort = __helpers.python.getDef("\n" + transformedCode, "merge_sort");
+    const { function_body } = merge_sort;
+
+    assert.match(function_body, /^\s{4}pass/m);
```
```suggestion
    assert.notMatch(function_body, /^\s{4}pass/m);
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6564aee9c077774ea49c3faf.md:
```diff
@@ -0,0 +1,53 @@
+
+The merge sort algorithm mainly performs three actions recursively: divide, conquer, and merge. Let's start with "divide".
+
+First, replace the `pass` keyword with a variable `mid_point`. Use the integer division operator (`//`) to divide the length of the `array` list by 2 and assign the result to your new `mid_point` variable. Remember to indent your code.
```
```suggestion
First, replace the `pass` keyword with a variable `middle_point`. Use the integer division operator (`//`) to divide the length of the `array` list by 2 and assign the result to your new `middle_point` variable. Remember to indent your code.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656639771fed09b5c0e8fe71.md:
````diff
@@ -0,0 +1,46 @@
+## --seed-contents--
+
+```py
+def merge_sort(array):
+    
+    middle_point = len(array) // 2
+    left_part = array[:middle_point]
+    right_part = array[middle_point:]
````
These lines must be inside the editable region. Please remove extra code block

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656627b47bd2d2a4afbc945d.md:
```diff
@@ -0,0 +1,57 @@
+
+# --hints--
+
+You should have a variable named `left_part`
```
```suggestion
You should have a variable named `left_part` in your `merge_sort` function.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656627b47bd2d2a4afbc945d.md:
````diff
@@ -0,0 +1,57 @@
+})
+```
+
+You should assign `array[middle_point:]` to the `left_part` variable
````
```suggestion
You should assign `array[:middle_point]` to the `left_part` variable
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656702f8b4cbd8cbf0a433c6.md:
```diff
@@ -0,0 +1,65 @@
+
+# --description--
+
+Create a `while` loop that compares an element in `left_part` to an element in `right_part`, and merges the smaller element to the main `array` list. 
```
```suggestion
Inside your function, create a `while` loop that compares an element in `left_part` to an element in `right_part`, and merges the smaller element to the main `array` list. 
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656706afd65547d22bee0b68.md:
````diff
@@ -0,0 +1,53 @@
+## --seed-contents--
+
+```py
+def merge_sort(array):
+    
+  middle_point = len(array) // 2
+  left_part = array[:middle_point]
+  right_part = array[middle_point:]
+
+  merge_sort(left_part)
+  merge_sort(right_part)
+
+  left_array_index = 0
+  right_array_index = 0
+  sorted_index = 0
+
+--fcc-editable-region--
+  while left_array_index < len(left_part) and right_array_index < len(right_part):
+    pass
+
+--fcc-editable-region--
````
```suggestion
def merge_sort(array):    
    middle_point = len(array) // 2
    left_part = array[:middle_point]
    right_part = array[middle_point:]

    merge_sort(left_part)
    merge_sort(right_part)

    left_array_index = 0
    right_array_index = 0
    sorted_index = 0
--fcc-editable-region--
    while left_array_index < len(left_part) and right_array_index < len(right_part):
        pass
--fcc-editable-region--
```

Dario-DC replied:
Fixed indentation

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656709e50ed928da35260276.md:
```diff
@@ -0,0 +1,57 @@
+
+# --description--
+
+In the previous step, you assigned the element in the `left_part` to the `array` list because it was smaller. But this will not always be the case. In some comparison cases, the element on the right could be smaller. 
```
```suggestion
In a previous step, you assigned the element in the `left_part` to the `array` list because it was smaller. But this will not always be the case. In some comparison cases, the element on the right could be smaller. 
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/65670d1ef177e7e2b76d9528.md:
```diff
@@ -0,0 +1,60 @@
+
+Each element assigned to the sorted array takes up an index in the list. So you have to move to the next index without an element.
+
+Outside the `if/else` block, increment `sorted_index` by 1. This should not be in the body of the `if` or `else` statement
```
```suggestion
Below the `if`/`else` block but still inside the `while` loop, increment `sorted_index` by 1. This should not be in the body of the `if` or `else` statement
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/65670d1ef177e7e2b76d9528.md:
```diff
@@ -0,0 +1,60 @@
+    right_array_index = 0
+    sorted_index = 0
+
+    while left_array_index < len(left_part) and right_array_index < len(right_part):
+        if left_part[left_array_index] < right_part[right_array_index]:
+            array[sorted_index] = left_part[left_array_index]
+            left_array_index += 1
+        else:
+            array[sorted_index] = right_part[right_array_index]
+            right_array_index += 1
+--fcc-editable-region--
+
+--fcc-editable-region--
```
```suggestion
--fcc-editable-region--
    while left_array_index < len(left_part) and right_array_index < len(right_part):
        if left_part[left_array_index] < right_part[right_array_index]:
            array[sorted_index] = left_part[left_array_index]
            left_array_index += 1
        else:
            array[sorted_index] = right_part[right_array_index]
            right_array_index += 1
--fcc-editable-region--
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/656710d1e0ec62253426db24.md:
```diff
@@ -0,0 +1,75 @@
+
+# --description--
+
+The first `while` loop we created compares one element from `left_part`with another in `right_part`, then adds the smaller element to the main `array` list.
```
```suggestion
The `while` loop you created compares one element from `left_part`with another in `right_part`, then adds the smaller element to the main `array` list.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569b5c820a6a1859786e774.md:
```diff
@@ -0,0 +1,78 @@
+
+# --description--
+
+Within the `while` loop, assign `right_part[right_array_index]` to `array[sorted_index]`
```
```suggestion
Within the `while` loop, assign `right_part[right_array_index]` to `array[sorted_index]`.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569b68fac723e8c20223ed3.md:
```diff
@@ -0,0 +1,65 @@
+
+# --description--
+
+Increment `right_array_index` by `1`.
```
```suggestion
Now, use the `+=` operator to increment `right_array_index` by `1`.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/657f59751d5a5c9b069ae0f3.md:
```diff
@@ -0,0 +1,76 @@
+
+You can use the `__name__` variable to determine if a Python script is being run as the main program or if it is being imported as a module (code written in another Python file).
+
+If the value of `__name__` is set to `'__main__'`, it implies that the current sciprt is the main program, and not a module.
```
```suggestion
If the value of `__name__` is set to `'__main__'`, it implies that the current script is the main program, and not a module.
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/657f59751d5a5c9b069ae0f3.md:
````diff
@@ -0,0 +1,76 @@
+Your `if` statement should check if `__name__` has a value of `'__main__'`
+
+```js
+assert.match(e.code.original["main.py"], /if\s+__name__\s*==\s*'__main__':/);
````
```suggestion
assert.match(e.code.original["main.py"], /if\s+__name__\s*==\s*("|')__main__\1\s*:/);
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/657f59751d5a5c9b069ae0f3.md:
````diff
@@ -0,0 +1,76 @@
+You should have the `pass` keyword in the body of your `if` statement
+
+```js
+assert.match(e.code.original["main.py"], /if\s+__name__\s*==\s*'__main__':\s*[^}]*\bpass\b/);
````
```suggestion
assert.match(e.code.original["main.py"], /if\s+__name__\s*==\s*("|')__main__\1\s*:\s*[^}]*\bpass\b/);
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c05b9166f9d5bb36c09e.md:
```diff
@@ -0,0 +1,63 @@
+
+# --hints--
+
+You should have `print("Unsorted array:")`
```
```suggestion
You should have `print('Unsorted array:')`
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c05b9166f9d5bb36c09e.md:
````diff
@@ -0,0 +1,63 @@
+You should have `print("Unsorted array:")`
+
+```js
+assert.match(code, e.code.original["main.py"], /^\s{4}print\s*\(\s*["']Unsorted array:["']\s*\)/m)
````
```suggestion
({ test: () => assert.match(e.code.original["main.py"], /^\s{4}print\s*\(\s*["']Unsorted array:["']\s*\)/m) })
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c1aeecaf95e25a3e2573.md:
````diff
@@ -0,0 +1,67 @@
+You should call the `merge_sort` function passing `numbers` as the argument.
+
+```js
+assert.match(code, /merge_sort\s*\(\s*numbers\s*\)/)
````
```suggestion
({ test: () => assert.match(code, /merge_sort\s*\(\s*numbers\s*\)/) })
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md:
````diff
@@ -0,0 +1,118 @@
+You should have `print("sorted array: " + str(numbers))` in your code. Remember to add the space after the colon
+
+```js
+assert.match(code, /print\("sorted numbers: " \+ str\(\s*numbers\s*\)\)/)
````
```suggestion
({ test: () => assert.match(e.code.original["main.py"], /^\s{4}print\s*\(\s*("|')Sorted array: \1\s*\+\s*str\s*\(\s*numbers\s*\)\s*\)/m) })
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md:
```diff
@@ -0,0 +1,118 @@
+
+# --hints--
+
+You should have `print("sorted array: " + str(numbers))` in your code. Remember to add the space after the colon
```
```suggestion
You should have `print('Sorted array: ' + str(numbers))` in your code. Remember to add the space after the colon
```

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md:
```diff
@@ -0,0 +1,118 @@
+
+At this point, the `numbers` list has been sorted. Call the `print` function to print string `Sorted array: ` and the sorted list.
+
+To do that, concatenate `"Sorted array: "` and `str(numbers)` in the `print` function.
```
```suggestion
To do that, concatenate `'Sorted array: '` and `str(numbers)` in the `print()` call.
```

Dario-DC reviewed (changes requested):
After this it should be ready to be merged. Except a couple of tests that are failing there shouldn't be other issue 👍

ihechikara committed:
update PR

Dario-DC commented on curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/655cd899f8de09431eabb40c.md:
```diff
@@ -0,0 +1,35 @@
+
+# --description--
+
+In this project, you'll learn data structures by building the merge sort algorithm. This is a sorting algorithm that uses the divide-and-conquer principle to sort collections of data. 
```
I would add a brief explanation of what divide and conquer means.

Dario-DC reviewed (commented):
It works fine for me. I left a few comments. Once they are resolved it should be god to go.

ihechikara committed:
update pr

ihechikara committed:
remove e.code.original and add intro

[skip ci]

Dario-DC reviewed (approved):
LGTM 🎉

moT01 committed:
fix: rebase and move to scientific computing

moT01 committed:
fix: order

moT01 committed:
fix: seed code for tests

moT01 committed:
Merge branch 'main' into feat/learn-data-structures-by-building-the-merge-sort-algorithm
